kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,48 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks.
+    /// This callback is called when a digest is computed.
+    /// The callback should follow the `Client` `hash_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes written on success
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8]) -> Result<usize, ReturnCode>;
+
+    /// Request the hardware block to generate a Digest
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self);
```

Comment:
In my proposed changes to the HIL I've changed the return type to `Result<(), ReturnCode>`, giving this the option to fail fast. It isn't stricly necessary, but can indicate the client that something's off without it having to wait on the callback.

Reply:
I'm expecting that the caller will be more interested in getting the result in the callback and that this wouldn't add a lot. I'm open to adding it though

Reply:
It doesn't add a lot and is fine this way. I started to think about what reasons could be making it fail fast. Those probably don't add up to the value of having this return `Result`.

Reply:
I added a result here